lib/spack/spack/solver/counter.py

```diff
@@ -199,109 +199,106 @@ def possible_dependencies(
             packages.append(current_spec.name)
 
         visited: Dict[str, Set[str]] = {}
-        for pkg_name in packages:
-            self._possible_dependencies(
-                pkg_name,
-                visited=visited,
-                transitive=transitive,
-                expand_virtuals=True,
-                depflag=allowed_deps,
-                virtuals=virtuals,
-                strict_depflag=strict_depflag,
-            )
+        self._possible_dependencies(
+            packages,
+            visited=visited,
+            transitive=transitive,
+            expand_virtuals=True,
+            depflag=allowed_deps,
+            virtuals=virtuals,
+            strict_depflag=strict_depflag,
+        )
 
         virtuals.update(self.runtime_virtuals)
         real_packages = set(visited) | self.runtime_pkgs
         return real_packages, virtuals
 
     def _possible_dependencies(
         self,
-        pkg_name: str,
+        packages: List[str],
         *,
         transitive: bool,
         expand_virtuals: bool = True,
         depflag: dt.DepFlag,
         strict_depflag: bool = False,
         visited: Optional[dict] = None,
-        missing: Optional[dict] = None,
         virtuals: set,
     ) -> Dict[str, Set[str]]:
+
+        stack = packages[:]
         visited = {} if visited is None else visited
-        missing = {} if missing is None else missing
-
-        visited.setdefault(pkg_name, set())
-
-        # At the moment exit early. Since libc is not buildable, there is no
-        # need to extend the search space with libc dependencies.
-        if pkg_name in self.context.libc_pkgs():
-            return visited
-
-        pkg_cls = self.context.repo.get_pkg_class(pkg_name=pkg_name)
-        for name, conditions in pkg_cls.dependencies_by_name(when=True).items():
-            if all(
-                self.context.unreachable(pkg_name=pkg_cls.name, when_spec=x) for x in conditions
-            ):
-                tty.debug(
-                    f"[{__name__}] Not adding {name} as a dep of {pkg_cls.name}, because "
-                    f"conditions cannot be met"
-                )
+
+        while stack:
+            pkg_name = stack.pop()
+
+            visited.setdefault(pkg_name, set())
```

Comment:
maybe just `if pkg_name in visited: continue; else visited[pkg_name] = set()`? then you dedupe the input too. I would keep `if dep_name in visited:` below too because the number of leaf nodes is of the order of the graph size.

Reply:
> Bit unclear why packages are followed but children are not.

I didn't change the implementation here, but I think this structure is there to deal with the `transitive=False` case. I'll try to simplify, or split, it.

Reply:
Done